raysect/optical/observer/sampler2d.pyx

```diff
@@ -27,30 +27,70 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
+import warnings
 import numpy as np
 from random import shuffle
 
 from raysect.optical.observer.base cimport FrameSampler2D
-from raysect.optical.observer.pipeline cimport RGBPipeline2D, RadiancePipeline2D, PowerPipeline2D
-from raysect.core.math cimport StatsArray1D, StatsArray2D, StatsArray3D
+from raysect.optical.observer.pipeline cimport RGBPipeline2D, RadiancePipeline2D, PowerPipeline2D, SpectralRadiancePipeline2D, SpectralPowerPipeline2D
+from raysect.core.math cimport StatsArray2D, StatsArray3D
 cimport numpy as np
 cimport cython
+ctypedef np.uint8_t uint8  # numpy boolean arrays are stored as 8-bit values
 
 
 cdef class FullFrameSampler2D(FrameSampler2D):
-    """ Evenly samples the full 2D frame. """
+    """
+    Evenly samples the full 2D frame or its masked fragment.
+
+    :param np.ndarray mask: The image mask array (default=None). A 2D boolean array with
+      the same shape as the frame. The tasks are generated only for those pixels for which
+      the mask is True.
+    """
+    cdef:
+        np.ndarray _mask
+        uint8[:, ::1] _mask_mv
+
+    def __init__(self, np.ndarray mask=None):
+
+        self.mask = mask
 
+    @property
+    def mask(self):
+        return self._mask
+
+    @mask.setter
+    def mask(self, np.ndarray value):
+        if value is None:
+            self._mask = None
+        else:
+            if value.ndim != 2:
+                raise ValueError("Mask must be a 2D array.")
+            self._mask = value.astype(np.bool)
+            self._mask_mv = np.frombuffer(self._mask, dtype=np.uint8).reshape(self.mask.shape)
```

Comment:
I've not run this yet, but won't this create a completely separate numpy array, only referenced by the memory view? i.e mask and mask_mv point to different objects? That would be a bad behaviour. Are you sure you can't use bool in the memory view?

Reply:
It might be better to promote the mask to integers internally and convert to bool externally if you can't obtain a bool from the memory view.

Reply:
For me, this was also a surprise, but numpy stores boolean values not as bits but as bytes (see the description of np.bool [here](https://docs.scipy.org/doc/numpy/user/basics.types.html)), so converting to np.uint8 does not create a new array. You can check it if you want:
```
a = np.ones(4, dtype=bool)
b = np.frombuffer(a, dtype=np.uint8).reshape(2, 2)
a[0] = False
print(b)
```